### PR TITLE
[BEAM-4858] Clean up division in batch size estimator.

### PR DESCRIPTION
**Pull request by robertwb**
In particular, this change

* Simplifies the aging out of old data. Not only was the old formula hard to
  understand, but it meant that bad data could stick around and poison the
  estimates forever.
* Adds a variance parameter allowing the batch size to vary over a fixed
  range giving a broader base for the linear regression.
* Uses numpy when available to do the regression.  This is both much more
  efficient and allows for less error-prone expression of more complicated
  analysis.

The algorithm was also changed to:

* Eliminates outliers, both using Cook's distance and just throwing out the
  top (often high-variance and high-influence) 20% when there is sufficient
  data.
* Weight by the inverse of batch size, to provide a more stable fixed size
  estimate (which the default "overhead" target is sensitive to).

These changes were tested against a large TFT pipeline and found to produce
more uniform batch sizes and similar, possibly slightly improved, total
runtimes and total costs.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by tvalentyn** [comment]
retest this please

**Comment by tvalentyn** [comment]
retest this please

**Comment by tvalentyn** [comment]
Looks like the Precommit failure is specific to this PR.

**Comment by robertwb** [comment]
Tests passing, PTAL.

File: sdks/python/apache_beam/transforms/util.py [hunk 1]
```diff
@@ -269,23 +270,68 @@ def record_time(self, batch_size):
...
-    pairs = sorted(zip(sorted_data[::2], sorted_data[1::2]),
-                   key=div_keys)
-    # Keep the top 1/3 most different pairs, average the top 2/3 most similar.
-    threshold = 2 * len(pairs) // 3
-    self._data = (
-        list(sum(pairs[threshold:], ()))
-        + [((x1 + x2) / 2.0, (t1 + t2) / 2.0)
-           for (x1, t1), (x2, t2) in pairs[:threshold]]
-        + odd_one_out)
+    # Make sure we don't change the parity of len(self._data).
```
Comment (tvalentyn) [inline sdks/python/apache_beam/transforms/util.py]: Why is this important?

File: sdks/python/apache_beam/transforms/util_test.py [hunk 2]
```diff
@@ -125,6 +126,75 @@ def test_target_overhead(self):
         clock.sleep(batch_duration(actual_sizes[-1]))
     self.assertEqual(expected_sizes, actual_sizes)
 
+  def test_variance(self):
+    clock = FakeClock()
```
Comment (tvalentyn) [inline sdks/python/apache_beam/transforms/util_test.py]: How about we fix the initial start time of FakeClock() to reduce potential for flakiness?

Comment (tvalentyn) [reply sdks/python/apache_beam/transforms/util_test.py]: Actually, it shouldn't matter.

File: sdks/python/apache_beam/transforms/util.py [hunk 3]
```diff
@@ -269,23 +270,68 @@ def record_time(self, batch_size):
...
+    self._data.pop(random.randrange(len(self._data) // 4))
+    self._data.pop(random.randrange(len(self._data) // 2))
+
+  @staticmethod
+  def linear_regression_no_numpy(xs, ys):
+    n = float(len(xs))
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/util.py]: Huh, I've never heard it called that before. Added an unambiguous comment.

File: sdks/python/apache_beam/transforms/util.py [hunk 4]
```diff
@@ -269,23 +270,68 @@ def record_time(self, batch_size):
...
+  @staticmethod
+  def linear_regression_numpy(xs, ys):
+    # pylint: disable=wrong-import-order, wrong-import-position
+    import numpy as np
+    from numpy import sum
+    xs = np.asarray(xs, dtype=float)
+    ys = np.asarray(ys, dtype=float)
+
+    # First do a simple least squares fit over all points.
+    n = len(xs)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/util.py]: Oh, that's a good one. It even supports weights (needed below). Done.

File: sdks/python/apache_beam/transforms/util.py [hunk 5]
```diff
@@ -269,23 +270,68 @@ def record_time(self, batch_size):
...
+    a = ybar - b * xbar
+
+    if n < 10:
+      return a, b
+    else:
+      # Refine this by throwing out outliers, according to Cook's distance.
+      sum_x2 = sum(xs**2)
+      errs = a * xs + b - ys
+      s2 = sum(errs**2) / (n - 2)
+      if s2 == 0:
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/util.py]: This is an absolute, not relative. measure of error, so there's not a default epsilon that makes sense. I simply want to divide by this below to normalize the error of each point to this "average" error, which is fine if it's tiny.

File: sdks/python/apache_beam/transforms/util.py [hunk 6]
```diff
@@ -269,23 +270,68 @@ def record_time(self, batch_size):
...
+        # It's an exact fit!
+        return a, b
+      h = (sum_x2 - 2 * sum_x * xs + n * xs**2) / (n * sum_x2 - sum_x**2)
+      cook_ds = 0.5 / s2 * errs**2 * (h / (1 - h)**2)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/util.py]: Added a link.

File: sdks/python/apache_beam/transforms/util.py [hunk 7]
```diff
@@ -316,17 +357,22 @@ def next_batch_size(self):
...
 
+    target = self._max_batch_size
+
     if self._target_batch_duration_secs:
       # Solution to a + b*x = self._target_batch_duration_secs.
-      cap = min(cap, (self._target_batch_duration_secs - a) / b)
+      target = min(target, (self._target_batch_duration_secs - a) / b)
 
     if self._target_batch_overhead:
       # Solution to a / (a + b*x) = self._target_batch_overhead.
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/util.py]: These are described in the public-facing docstring below. I plumbed variance through there as well now.

File: sdks/python/apache_beam/transforms/util.py [hunk 8]
```diff
@@ -316,17 +357,22 @@ def next_batch_size(self):
...
+      target = min(target, (a / b) * (1 / self._target_batch_overhead - 1))
 
-    # Avoid getting stuck at min_batch_size.
+    # Avoid getting stuck.
     jitter = len(self._data) % 2
-    return int(max(self._min_batch_size + jitter, cap))
+    if len(self._data) > 10:
+      target += int(target * self._variance * 2 * (random.random() - .5))
+
+    return int(max(self._min_batch_size + jitter, min(target, cap)))
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/util.py]: Clarified.

File: sdks/python/apache_beam/transforms/util.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tvalentyn) [reply sdks/python/apache_beam/transforms/util.py]: Thanks, cross-checked :-). Did you also compute h by pen-and-paper or there is a faster way?

File: sdks/python/apache_beam/transforms/util.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/util.py]: Right. Done.

File: sdks/python/apache_beam/transforms/util.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/util.py]: Clarified. The choice of approximation isn't important here (and I don't want to give the impression that it's a non-linear polynomial fit.)

File: sdks/python/apache_beam/transforms/util.py [hunk 9]
```diff
@@ -269,23 +270,60 @@ def record_time(self, batch_size):
...
+    n = len(xs)
+    if n < 10:
+      return a, b
+    else:
+      # Refine this by throwing out outliers, according to Cook's distance.
+      # https://en.wikipedia.org/wiki/Cook%27s_distance
+      sum_x = sum(xs)
+      sum_x2 = sum(xs**2)
+      errs = a * xs + b - ys
+      s2 = sum(errs**2) / (n - 2)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/util.py]: The typical definition takes into account the number of degrees of freedom.

File: sdks/python/apache_beam/transforms/util.py [hunk 10]
```diff
@@ -269,23 +270,60 @@ def record_time(self, batch_size):
...
+      if s2 == 0:
+        # It's an exact fit!
+        return a, b
+      h = (sum_x2 - 2 * sum_x * xs + n * xs**2) / (n * sum_x2 - sum_x**2)
+      cook_ds = 0.5 / s2 * errs**2 * (h / (1 - h)**2)
+
+      # Re-compute the regression, excluding those points with Cook's distance
+      # greater than 1.
+      b, a = np.polyfit(xs, ys, 1, w=cook_ds < 1)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/transforms/util.py]: Reconciled.

**Comment by tvalentyn** [comment]
Thanks. The code LGTM, PTAL at test failures and please do a performance check using updated code before merging.

**Comment by robertwb** [comment]
You're right, the a and b were switched in computing the error term when I copied this to the PR. This meant that significantly more points were considered outliers (but enough retained to typically give a reasonable regression). Unfortunately this fix means that it's still pretty sensitive to multiple outliers...

I'm trying a simpler approach: just assume the top quantile is outliers. We have enough data to make this pretty robust. Running experiments now. 

(As for computing h, I used sagemath.)

**Comment by tvalentyn** [comment]
Thanks. Is this ready for review?

**Comment by robertwb** [comment]
Yes, this is ready for another round of review. I also cleaned up the commit history (with an exanded and updated commit comment.)

File: sdks/python/apache_beam/transforms/util.py [hunk 11]
```diff
@@ -269,23 +270,62 @@ def record_time(self, batch_size):
...
+      if s2 == 0:
+        # It's an exact fit!
+        return a, b
+      h = (sum_x2 - 2 * sum_x * xs + n * xs**2) / (n * sum_x2 - sum_x**2)
+      cook_ds = 0.5 / s2 * errs**2 * (h / (1 - h)**2)
+
+      # Re-compute the regression, excluding those points with Cook's distance
+      # greater than 0.5, and weighting by the inverse of x to give a more
+      # stable y-intercept.
+      weight = (cook_ds <= 0.5) / xs
```
Comment (tvalentyn) [inline sdks/python/apache_beam/transforms/util.py]: What is the intuition behind weighing by the inverse of x? Do we assume that we have to process smaller batches more often than larger batches, and therefore smaller batches datapoints should receive more consideration than larger batches? Why is the instability caused by larger xs, and not by smaller?

Comment (tvalentyn) [reply sdks/python/apache_beam/transforms/util.py]: I see, thanks for taking time to explain on PR & comments.

**Pull request merged**